Ensure COM library is active when D2D renderer is in use.

a-wi opened the pull request:
Modify the order of cleaning up the application in order to assure that modules are cleaned up when wxThpApp exists and its resources are still available.
See http://trac.wxwidgets.org/ticket/17308

vadz commented:
Unfortunately I don't think this is safe to do because the (_user_-defined) `wxApp`-derived class dtor could rely on the functionality provided by some modules to still be available. In fact, while I don't remember the details, I think we have actually changed the cleanup order exactly in order to avoid this problem.

I think a better fix for http://trac.wxwidgets.org/ticket/17308 would be to refactor COM initialization in a module of its own and make the D2D one dependent on it, what do you think?

a-wi committed:
Ensure COM library is active when Direct2D renderer is in use.

Some internal D2D operations need active COM library so ensure it is properly initialized and released in wxDirect2DModule.

a-wi commented:
The concept of self-contained modules is very good so in the current patch D2D module takes care of COM initialization/releasing on its own.

Regarding life cycle of the modules: This asymmetry between initialization and clean up sequence looks to me a bit confusing and counter-intuitive. If modules need to be cleaned up after the application object so maybe they could be initialized before it (not after as it is done now)?

vadz commented:
I think it might be better to initialize the modules before creating `wxTheApp`, but we need to check if no modules rely on something currently being initialized by the app (notably in its overridden `OnInitGui()` method).

In any case, I don't have any problems with the current patch and will merge it after the CI builds finish (because it seems like the crashes in the tests for the initial version did indicate a real problem after all), thanks!

vadz commented:
Committed as 450019dc218f94146a30549994976e9d804b1c05, thanks!
